vibing-ai/marvel-ai-backend
# Note generator app

## Description

I developed a Notes Generator Tool to generate concise, structured notes from various types of input in other to provide students with clear, summarized learning materials. 

## Related Issue

[https://github.com/orgs/marvelai-org/projects/8/views/1?pane=issue&itemId=89621248&issue=marvelai-org%7Cmarvel-ai-backend%7C129](url)

## Type of Change

- [ ] **New feature**: A non-breaking change that adds "Note generator" feature to the app.
- [ ] **Documentation update**: New feature code is well documented

## Proposed Solution

1. Input Handling
    * Implementation:
        * Built logic to accept three types of input: plain text, file uploads (PLAIN TEXT, CSV, PDF, DOCX), and URLs (websites, YouTube.
       * Used helper functions (extract_text_from_file, extract_text_from_url) for modularity, making the code extensible for future input types.
    * Design Decisions:
        * Modularized input parsing to support easy integration of new file formats.
        * Used mock-based testing to isolate file and URL handlers.
2. Notes Generation Logic
    * Implementation:
        * Developed an executor function that coordinates input parsing, focus-based filtering, and AI-driven note summarization.
        * Used flexible output structures (bullet points, paragraphs, tables) to match user preferences.
    * Design Decisions:
        * Separated content preprocessing and notes generation for clarity and maintainability.
        * Added an optional "focus" parameter to tailor notes to specific topics, improving usability.
3. Output Formats
    * Implementation:
        * Created functions to export notes in plain text, DOCX, and PDF formats using libraries like docx and FPDF.
   * Design Decisions:
        * Chose widely used formats to ensure compatibility with users' needs.
        * Structured output generation modularly to enable quick addition of new export formats.
4. Core Architecture
    * Implementation:
        * Organized code into a structured folder hierarchy (app/features/notes_generator/) for maintainability.
       * Used core.py for primary logic and tools.py for reusable helper functions.
    * Design Decisions:
        * Followed a clean architecture with clear separation of responsibilities to improve scalability and readability.
        * Emphasized reusability with helper abstractions.
5. Testing
    * Implementation:
        * Created unit tests (test_core.py, test_tools.py) for all primary and helper functions using PyTest.
        * Covered normal operations, edge cases, and error handling.
    * Design Decisions:
        * Prioritized test-driven development (TDD) to ensure code reliability.
        * Mocked dependencies for isolated testing and ensured 100% coverage for core functionalities.
6. Logging and Error Handling
    * Implementation:
        * Added comprehensive logging to track errors and user inputs for debugging.
    * Design Decisions:
        * Followed platform standards for consistent and meaningful error reporting.
        * Prioritized user-friendly error messages to enhance user experience.
7. Extensibility and Maintainability
    * Implementation:
       * Used metadata files (metadata.json) to define input parameters for validation.
        * Documented all functions with type hints and docstrings.
    * Design Decisions:
        * Focused on future-proofing by using modular and extensible code structures.
       * Prioritized readability with consistent naming conventions and clear comments.

## How to Test

 Test Configurations
1. Ensure you have the necessary dependencies installed: pip install -r requirements.txt 
2. Run the tests: PYTHONPATH=./ pytest app/features/notes_generator/tests/test_core.py

Test Cases and Expected Outcomes
1. test_executor_text_input_valid:
    * Description: Tests the executor function with valid text input.
    * Expected Outcome: The function should return a status of "success". 
2. test_executor_file_input_valid:
    * Description: Tests the executor function with valid file input.
    * Expected Outcome: The function should return a status of "success". 
3. test_executor_invalid_file_type:
    * Description: Tests the executor function with an unsupported file type.
    * Expected Outcome: The function should raise a ValueError.
4. test_extract_text_from_file:
    * Description: Tests the extract_text_from_file function with a valid file.
    * Expected Outcome: The function should return the extracted text.
5. test_extract_text_from_url:
    * Description: Tests the extract_text_from_url function with a valid URL.
    * Expected Outcome: The function should return the extracted text.

## Unit Tests

1. test_executor_text_input_valid:
    * Tests the executor function with valid text input.
2. test_executor_file_input_valid:
    * Tests the executor function with valid file input.
3. test_executor_invalid_file_type:
    * Tests the executor function with an unsupported file type.
4. test_extract_text_from_file:
    * Tests the executor function with dummy.pdf file.
6. test_extract_text_from_url:
    * Tests the executor function with valid YouTube video ID

## Documentation Updates

N/A



## Additional Information

I am look forward to your review and feedback
